src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,694 @@
+    if (!zookeeper_client || zookeeper_client->expired())
+    {
+        zookeeper_client = nullptr;
+        if (zookeeper_name == "default")
+            zookeeper_client = getContext()->getZooKeeper();
+        else
+            zookeeper_client = getContext()->getAuxiliaryZooKeeper(zookeeper_name);
+
+        zookeeper_client->sync(rootKeeperPath());
+    }
+
+    return zookeeper_client;
+}
+
+const std::string & StorageKeeperMap::rootKeeperPath() const
+{
+    return root_path;
+}
+
+std::string StorageKeeperMap::fullPathForKey(const std::string_view key) const
+{
+    return fmt::format("{}/{}", root_path, key);
+}
+
+UInt64 StorageKeeperMap::keysLimit() const
+{
+    return keys_limit;
+}
+
+std::optional<bool> StorageKeeperMap::isTableValid() const
+{
+    std::lock_guard lock{init_mutex};
+    if (table_is_valid.has_value())
+        return *table_is_valid;
+
+    try
+    {
+        // validate all metadata nodes are present
+        Coordination::Requests requests;
+        requests.push_back(zkutil::makeCheckRequest(table_path, -1));
+
+        Coordination::Responses responses;
+        auto client = getClient();
+        auto res = client->tryMulti(requests, responses);
+        table_is_valid = res == Coordination::Error::ZOK;
+    }
+    catch (const Coordination::Exception & e)
+    {
+        tryLogCurrentException(log);
+
+        if (!Coordination::isHardwareError(e.code))
+            table_is_valid = false;
+    }
+    catch (const Exception &)
+    {
+        tryLogCurrentException(log);
+        table_is_valid = false;
+    }
+
+    return table_is_valid;
+}
+
+Chunk StorageKeeperMap::getByKeys(const ColumnsWithTypeAndName & keys, PaddedPODArray<UInt8> & null_map, const Names &) const
+{
+    if (keys.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "StorageKeeperMap supports only one key, got: {}", keys.size());
+
+    auto raw_keys = serializeKeysToRawString(keys[0]);
+
+    if (raw_keys.size() != keys[0].column->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Assertion failed: {} != {}", raw_keys.size(), keys[0].column->size());
+
+    return getBySerializedKeys(raw_keys, &null_map);
+}
+
+Chunk StorageKeeperMap::getBySerializedKeys(const std::span<const std::string> keys, PaddedPODArray<UInt8> * null_map) const
+{
+    Block sample_block = getInMemoryMetadataPtr()->getSampleBlock();
+    MutableColumns columns = sample_block.cloneEmptyColumns();
+    size_t primary_key_pos = getPrimaryKeyPos(sample_block, getPrimaryKey());
+
+    if (null_map)
+    {
+        null_map->clear();
+        null_map->resize_fill(keys.size(), 1);
+    }
+
+    auto client = getClient();
+
+    std::vector<std::future<Coordination::GetResponse>> values;
+    values.reserve(keys.size());
+
+    for (const auto & key : keys)
+    {
+        const auto full_path = fullPathForKey(key);
+        if (full_path == metadata_path)
+        {
+            values.emplace_back();
+            continue;
```

Comment:
I think `if (key.empty())` will be more readable

Reply:
what do you mean? Why would the key be empty?

Reply:
we get all the children for select query without filter, and to avoid removing a member of a vector I decided to check here while iterating if one of the keys is the metadata node.

Reply:
It should not be possible to pass `metadata` as key